v21.2/bulk-delete-data.md

```diff
@@ -140,25 +140,71 @@ CockroachDB records the timestamp of each row created in a table in the `crdb_in
 
 CockroachDB does not support Time to Live (TTL) on table rows. To delete "expired" rows, we recommend automating a batch delete process using a job scheduler like `cron`.
 
-For example, suppose that every morning you want to delete all rows in the [`tpcc`](cockroach-workload.html#tpcc-workload) `history` table that are older than a month. To do this, you could use the example Python script that [batch deletes on the non-indexed `h_date` column](#batch-delete-on-a-non-indexed-column).
+For example, suppose that every morning you want to delete all rows in the [`rides` table](movr.html#the-movr-database) in the [`movr` database](movr.html) that are older than a month. To do this, you could write a Python script that batch-deletes rows based on the values of an indexed [`TIMESTAMP`](timestamp.html) column, and then run the script with a daily `cron` job.
 
-To run the script with a daily `cron` job:
+1. To record the last day and time a row was updated, create a `TIMESTAMP` column with an [`ON UPDATE` expression](add-column.html):
+
+    {% include copy-clipboard.html %}
+    ~~~ sql
+    USE movr;
+    ALTER TABLE rides ADD COLUMN last_updated TIMESTAMPTZ DEFAULT now() ON UPDATE now();
+    ~~~
+
+1. To improve `DELETE` performance, index the `TIMESTAMP` column. We recommend using a [hash-sharded index](hash-sharded-indexes.html) to reduce bottlenecks, as `TIMESTAMP` values are [sequentially stored in ranges](architecture/distribution-layer.html#range-splits):
+
+    {% include copy-clipboard.html %}
+    ~~~ sql
+    SET experimental_enable_hash_sharded_indexes=on;
+    CREATE INDEX ON rides(last_updated) USING HASH WITH BUCKET_COUNT=8;
+    ~~~
+
+1. Write a script with a batch-delete loop, following the [Batch delete on an indexed column pattern](#batch-delete-on-an-indexed-column):
+
+    {% include copy-clipboard.html %}
+    ~~~ python
+    #!/usr/bin/env python3
+
+    import psycopg2
+    import psycopg2.sql
+    import os
+    import datetime
+
+    conn = psycopg2.connect(os.environ.get('DB_URI'))
+    filter = datetime.datetime.utcnow() - datetime.timedelta(days=30)
+    lastrow = None
+
+    while True:
+      with conn:
+        with conn.cursor() as cur:
+            if lastrow:
+                filter = lastrow[0]
+            query = psycopg2.sql.SQL("DELETE FROM rides WHERE last_updated < %s ORDER BY last_updated DESC LIMIT 5000 RETURNING last_updated")
+            cur.execute(query, (filter,))
```

Comment:
i think it would be more idiomatic to do
```
cur.execute("DELETE FROM rides WHERE last_updated < %s ORDER BY last_updated DESC LIMIT 5000 RETURNING last_updated", [filter])
```

the `psycopg2.sql` is useful for when you have to escape identifiers like table names

Reply:
i highly record `SELECT <pk> FROM rides AS OF SYSTEM TIME `-30s` WHERE ....`, and then running `BEGIN LOW PRIORITY TRANSACTION; DELETE ....; COMMIT`

Reply:
@rafiss 

> i think it would be more idiomatic to do
> 
> ```
> cur.execute("DELETE FROM rides WHERE last_updated < %s ORDER BY last_updated DESC LIMIT 5000 RETURNING last_updated", [filter])
> ```
> 
> the `psycopg2.sql` is useful for when you have to escape identifiers like table names

Done

Reply:
@otan 

> i highly record `SELECT <pk> FROM rides AS OF SYSTEM TIME `-30s` WHERE ....`, and then running `BEGIN LOW PRIORITY TRANSACTION; DELETE ....; COMMIT`

I'm a lil' confused. Are you suggesting a manual transaction in the example here, or just explaining your TTL prototype?

Reply:
im suggesting this example will cause performance issues in production based on my testing of what this webpage recommend, and to avoid that, one should do

```
BEGIN
SELECT <pk> FROM rides AS OF SYSTEM TIME -30s WHERE <delete criiteria>
ROLLBACK

BEGIN LOW PRIORITY TRANSACTION; DELETE FROM rides WHERE PK IN (...); COMMIT
```

Reply:
Ahh. I see. Perhaps we merge this PR so we can have updated TTL workaround guidance with `ON UPDATE`, and then I make a new issue about the performance improvement based on your recommendation?